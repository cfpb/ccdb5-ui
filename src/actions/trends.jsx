import { REQUERY_ALWAYS, REQUERY_NEVER } from '../constants'

export const CHART_TYPE_CHANGED = 'CHART_TYPE_CHANGED'
export const DATA_LENS_CHANGED = 'DATA_LENS_CHANGED'
export const DATA_SUBLENS_CHANGED = 'DATA_SUBLENS_CHANGED'
export const DEPTH_CHANGED = 'DEPTH_CHANGED'
export const DEPTH_RESET = 'DEPTH_RESET'
export const FOCUS_CHANGED = 'FOCUS_CHANGED'
export const FOCUS_REMOVED = 'FOCUS_REMOVED'
export const TRENDS_TOOLTIP_CHANGED = 'TRENDS_TOOLTIP_CHANGED'

/**
 * Notifies the application that chart type toggled
 *
 * @param {string} chartType which chartType was selected, line or stacked area
 * @returns {string} a packaged payload to be used by Redux reducers
 */
export function changeChartType( chartType ) {
  return {
    type: CHART_TYPE_CHANGED,
    chartType,
    requery: REQUERY_NEVER
  }
}

/**
 * Notifies the application that data lens overview, product, issue was toggled
 *
 * @param {string} lens which lens was selected
 * @returns {string} a packaged payload to be used by Redux reducers
 */
export function changeDataLens( lens ) {
  return {
    type: DATA_LENS_CHANGED,
    lens,
    requery: REQUERY_ALWAYS
  }
}

/**
 * Indicates the data subLens selected
 *
 * @param {string} subLens the tab selected for row charts
 * @returns {string} a packaged payload to be used by Redux reducers
 */
export function changeDataSubLens( subLens ) {
  return {
    type: DATA_SUBLENS_CHANGED,
    requery: REQUERY_ALWAYS,
    subLens
  }
}

/**
 * Notifies the application that depth is being changed
 *
 * @param {string} depth the max number of aggregations returned
 * @returns {string} a packaged payload to be used by Redux reducers
 */
export function changeDepth( depth ) {
  return {
    type: DEPTH_CHANGED,
    requery: REQUERY_ALWAYS,
    depth
  }
}

/**
 * Notifies the application that depth is being reset
 *
 * @returns {string} a packaged payload to be used by Redux reducers
 */
export function resetDepth() {
  return {
    type: DEPTH_RESET,
    requery: REQUERY_ALWAYS
  }
}

/**
 * Notifies the application that focus is being changed
 *
 * @param {string} focus the text to search for
 * @param {string} lens the lens we're focusing on
 * @param {array} filterValues the parent/child focus sub-aggs to apply
 * @returns {string} a packaged payload to be used by Redux reducers
 */
export function changeFocus( focus, lens, filterValues ) {
  return {
    type: FOCUS_CHANGED,
    requery: REQUERY_ALWAYS,
    filterValues,
    focus,
    lens
  }
}

/**
 * Notifies the application that focus is being removed
 *
 * @returns {string} a packaged payload to be used by Redux reducers
 */
export function removeFocus() {
  return {
    type: FOCUS_REMOVED,
    requery: REQUERY_ALWAYS
  }
}


/**
 * Notifies the application that the toolTip for stacked area chart has changed
 *
 * @param {string} value the new payload from the tooltip
 * @returns {string} a packaged payload to be used by Redux reducers
 */
export function updateTrendsTooltip( value ) {
  return {
    type: TRENDS_TOOLTIP_CHANGED,
    value,
    requery: REQUERY_NEVER
  }
}
