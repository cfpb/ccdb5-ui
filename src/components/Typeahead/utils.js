/**
 * Sets the options for the async typeahead
 *
 * @param {string} value - value typed into the typeahead
 * @param {Function} setOptions - function to set local array of options
 * @param {string} uri - the endpoint being hit by the fetch
 * @returns {void | Promise<object>} - Either nothing or a response with option object
 */
export function handleFetchSearch(value, setOptions, uri) {
  const num = value.toLowerCase();
  if (num === '') {
    setOptions([]);
    return;
  }
  return fetch(uri)
    .then((result) => result.json())
    .then((items) => {
      const options = items.map((item) => ({
        key: item,
        label: item,
        position: item.toLowerCase().indexOf(num),
        value,
      }));
      setOptions(options);
    });
}
