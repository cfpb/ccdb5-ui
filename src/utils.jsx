import { DATE_RANGE_MIN, SLUG_SEPARATOR } from './constants'
import moment from 'moment'

// eslint-disable-next-line complexity
export const calculateDateInterval = ( minDate, maxDate ) => {
  // only check intervals if the end date is today
  // round off the date so the partial times don't mess up calculations
  const today = moment( new Date() ).startOf( 'day' )
  const end = moment( maxDate ).startOf( 'day' )
  const start = moment( minDate ).startOf( 'day' )

  // make sure end date is the same as today's date
  if ( end.diff( today, 'days', true ) !== 0 ) {
    return ''
  }

  // is the start date the same as the oldest document?
  if ( moment( minDate ).isSame( DATE_RANGE_MIN, 'day' ) ) {
    return 'All'
  }

  // verify if it's 3 or 1 years
  const yrDiff = end.diff( start, 'years', true )

  if ( yrDiff === 3 || yrDiff === 1 ) {
    return yrDiff + 'y'
  }

  // verify if it's 6 or 3 months
  const moDiff = end.diff( start, 'months', true )
  if ( moDiff === 6 || moDiff === 3 ) {
    return moDiff + 'm'
  }

  return ''
}

/**
 * Function to set the limit of the range of a set of numbers
 * @param {int} x value we are checking
 * @param {int} min smallest number it can me
 * @param {int} max biggest number it can be
 * @returns {*}the limited value
 */
export const clamp = ( x, min, max ) => {
  if ( x < min ) {
    x = min;
  } else if ( x > max ) {
    x = max;
  }
  return x;
}

/**
 * Replacement for the common pattern:
 * if( o.field )
 *    x = o.field
 * else
 *    x = alternateValue
 *
 * Avoids some of the complexity lint warnings
 *
 * @param {Object} o the object being tested
 * @param {string} field the field to check
 * @param {string|Object} alternateValue the value to use in absence
 * @returns {string} the value to use
 */
export const coalesce = ( o, field, alternateValue ) => {
  if ( typeof o !== 'object' ) {
    return alternateValue;
  }

  return field in o && o[field] ? o[field] : alternateValue;
};

/**
 * Creates a hash from a string
 *
 * @param {string} someString the string to hash
 * @returns {int} a hashing of the string
 */
export function hashCode( someString ) {
  const s = String( someString )
  let hash = 0
  let i, chr
  if ( s.length === 0 ) {
    return hash
  }
  for ( i = 0; i < s.length; i++ ) {
    chr = s.charCodeAt( i )
    hash = ( hash << 5 ) - hash + chr

    // Convert to 32bit integer
    hash |= 0
  }
  return hash
}

/**
 * Creates a hash from an object
 *
 * @param {string} o the object to hash
 * @returns {string} a hashing of the object
 */
export function hashObject( o ) {
  return hashCode( JSON.stringify( o ) )
}
export const normalize = s => s.toLowerCase()

export const slugify = ( a, b ) => a + SLUG_SEPARATOR + b

export const sortSelThenCount = ( options, selected ) => {
  const retVal = ( options || [] ).slice()

  // Sort the array so that selected items appear first, then by doc_count
  /* eslint complexity: ["error", 5] */
  retVal.sort( ( a, b ) => {
    const aSel = selected.indexOf( a.key ) !== -1
    const bSel = selected.indexOf( b.key ) !== -1

    if ( aSel && !bSel ) {
      return -1;
    }
    if ( !aSel && bSel ) {
      return 1;
    }

    // Both are selected or not selected
    // Sort by descending doc_count
    return b.doc_count - a.doc_count
  } )

  return retVal
}

/**
* Safely format a date
*
* @param {Date} date the date to convert
* @returns {string} the date formatted for the current locale
*/
export function shortFormat( date ) {
  const wrapped = moment( date )
  return date ? wrapped.format( 'M/D/YYYY' ) : ''
}

/**
* Convert a date to a truncated ISO-8601 string
*
* @param {Date} date the date to convert
* @returns {string} the date formatted as yyyy-mm-ddd
*/
export function shortIsoFormat( date ) {
  return date ? date.toISOString().substring( 0, 10 ) : ''
}

// ----------------------------------------------------------------------------
// attribution: underscore.js (MIT License)

/**
* Delay the implementation of a function until after a period of time
* This prevents expensive calls from being made while triggering events are
* still happening
*
* @param {function} func a function with an embedded expensive call
* @param {int} wait the number of msecs to delay before calling the function
* @returns {function} a replacement function to use in place of the original
*/
export function debounce( func, wait ) {
  var timer = null;

  var later = ( context, args ) => {
    timer = null;
    func.apply( context, args );
  }

  return ( ...theArgs ) => {
    if ( !timer ) {
      timer = setTimeout( later, wait, this, theArgs )
    }
  }
}

// ----------------------------------------------------------------------------
// attribution: lodash.js (Creative Commons License)

/**
* Binds methods of an object to the object itself, overwriting the existing
* method
*
* @param {Object} obj The object to bind and assign the bound methods to.
* @param {...(string|string[])} methodNames The object method names to bind,
*  specified individually or in arrays.
* @returns {Object} the updated object
*/
export function bindAll( obj, methodNames ) {
  const length = methodNames.length
  for ( let i = 0; i < length; i++ ) {
    const key = methodNames[i]
    obj[key] = obj[key].bind( obj )
  }
  return obj;
}
