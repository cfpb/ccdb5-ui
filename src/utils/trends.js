import * as types from '../constants';
import dayjs from 'dayjs';
// ----------------------------------------------------------------------------
export const showCompanyOverLay = (lens, companyFilters, isLoading) => {
  if (isLoading) {
    return false;
  }

  // we need to show the companyOverlay if:
  // lens === 'Company' AND there are no company filters
  if (lens === 'Company') {
    return !companyFilters || companyFilters.length === 0;
  }

  return false;
};

export const getSubLens = (lens) => {
  if (!lens) {
    return '';
  }

  switch (lens) {
    case 'Overview':
      return '';
    case 'Company':
      return 'product';
    default:
      return 'sub_' + lens.toLowerCase();
  }
};

/**
 * helper function to strip out the "Other" data points from stacked area if
 * they don't have any values
 *
 * @param {Array} buckets - contains all of the date points for the stacked area
 * @returns {Array} cleaned up array or not
 */
export const pruneOther = (buckets) => {
  const sumOther = buckets
    .filter((bucket) => bucket.name === 'Other')
    .reduce((prev, cur) => prev + cur.value, 0);

  return sumOther > 0
    ? buckets
    : buckets.filter((bucket) => bucket.name !== 'Other');
};

export const isGreaterThanYear = (from, to) => {
  const fromDay = dayjs(from);
  const toDay = dayjs(to);
  return toDay.diff(fromDay, 'days') > 366;
};

/**
 * gets the valid intervals based on dates selected.
 * get rid of Day if range > 1yr
 *
 * @param {object} from - date
 * @param {object} to - date
 * @returns {Array} array of date intervals
 */
export const getIntervals = (from, to) =>
  types.dateIntervals.map((interval) => ({
    name: interval,
    disabled: isGreaterThanYear(from, to) && interval === 'Day',
  }));

/**
 * trigger this after a user clicks a focus.  we scroll to the select box
 * so that users don't get a blank wall of content
 */
export const scrollToFocus = () => {
  const lensSelect = document.getElementById('search-summary');
  if (lensSelect) {
    lensSelect.scrollIntoView();
  }
};
