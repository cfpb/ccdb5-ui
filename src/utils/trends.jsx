import * as types from '../constants'
import dayjs from 'dayjs'

// ----------------------------------------------------------------------------
export const showCompanyOverLay = ( lens, companyFilters, isLoading ) => {
  if ( isLoading ) {
    return false
  }

  // we need to show the companyOverlay if:
  // lens === 'Company' AND there are no company filters
  if ( lens === 'Company' ) {
    return !companyFilters || companyFilters.length === 0
  }

  return false
}

/* eslint-disable-next-line no-extra-parens */
export const getSubLens = lens => {
  if ( !lens ) {
    return ''
  }

  switch ( lens ) {
    case 'Overview':
      return ''
    case 'Company':
      return 'product'
    default:
      return 'sub_' + lens.toLowerCase()
  }
}

/**
 * helper function to strip out the "Other" data points from stacked area if
 * they don't have any values
 * @param {array} buckets contains all of the date points for the stacked area
 * @returns {array} cleaned up array or not
 */
export const pruneOther = buckets => {
  const sumOther = buckets
    .filter( o => o.name === 'Other' )
    .reduce( ( prev, cur ) => prev + cur.value, 0 )

  return sumOther > 0 ? buckets : buckets.filter( o => o.name !== 'Other' )
}

export const isGreaterThanYear = ( from, to ) => {
  const a = dayjs( from )
  const b = dayjs( to )
  return b.diff( a, 'days' ) > 366
}

/**
 * gets the valid intervals based on dates selected.
 * get rid of Day if range > 1yr
 * @param {object} from date
 * @param {object} to date
 * @returns {array} array of date intervals
 */
export const getIntervals = ( from, to ) =>
  types.dateIntervals.map( o => ( {
    name: o,
    disabled: isGreaterThanYear( from, to ) && o === 'Day'
  } ) )

/**
 * trigger this after a user clicks a focus.  we scroll to the select box
 * so that users don't get a blank wall of content
 */
export const scrollToFocus = () => {
  const lensSelect = document.getElementById( 'search-summary' )
  if ( lensSelect ) {
    lensSelect.scrollIntoView()
  }
}
