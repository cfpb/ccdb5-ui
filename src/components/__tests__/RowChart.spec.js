import * as trendsUtils from '../../utils/trends';
import configureMockStore from 'redux-mock-store';
import ReduxRowChart, {
  mapDispatchToProps,
  mapStateToProps,
  RowChart,
} from '../Charts/RowChart';
import { mount, shallow } from 'enzyme';
import { Provider } from 'react-redux';
import React from 'react';
import renderer from 'react-test-renderer';
import thunk from 'redux-thunk';
import * as utils from '../../utils';

// this is how you override and mock an imported constructor
jest.mock('britecharts', () => {
  const props = [
    'row',
    'margin',
    'backgroundColor',
    'colorSchema',
    'enableLabels',
    'labelsSize',
    'labelsTotalCount',
    'labelsNumberFormat',
    'outerPadding',
    'percentageAxisToMaxRatio',
    'yAxisLineWrapLimit',
    'miniTooltip',
    'yAxisPaddingBetweenChart',
    'width',
    'wrapLabels',
    'height',
    'on',
    'valueFormatter',
    'paddingBetweenGroups',
  ];

  const mock = {};

  for (let idx = 0; idx < props.length; idx++) {
    const propName = props[idx];
    mock[propName] = jest.fn().mockImplementation(() => {
      return mock;
    });
  }

  return mock;
});

jest.mock('d3', () => {
  const props = [
    'select',
    'each',
    'node',
    'getBoundingClientRect',
    'width',
    'datum',
    'call',
    'remove',
    'selectAll',
    'on',
  ];

  const mock = {};

  for (let idx = 0; idx < props.length; idx++) {
    const propName = props[idx];
    mock[propName] = jest.fn().mockImplementation(() => {
      return mock;
    });
  }

  // set narrow width value for 100% test coverage
  mock.width = 100;

  return mock;
});

/**
 * @returns {void}
 */
function setupSnapshot() {
  const middlewares = [thunk];
  const mockStore = configureMockStore(middlewares);
  const store = mockStore({
    aggs: {},
    query: {
      lens: 'bar',
      tab: 'foo',
    },
    view: {
      expandedRows: [],
      isPrintMode: false,
      width: 1000,
    },
  });

  return renderer.create(
    <Provider store={store}>
      <ReduxRowChart
        id="foo"
        data={[1, 2, 3]}
        title="Foo title we want"
        helperText="sub title under title"
        colorScheme={[]}
        total={1000}
      />
    </Provider>,
  );
}

describe('component: RowChart', () => {
  describe('initial state', () => {
    xit('renders without crashing', () => {
      const target = setupSnapshot();
      const tree = target.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('componentDidUpdate', () => {
    let mapDiv;

    beforeEach(() => {
      mapDiv = document.createElement('div');
      mapDiv.setAttribute('id', 'row-chart-foo');
      window.domNode = mapDiv;
      document.body.appendChild(mapDiv);
    });

    afterEach(() => {
      const div = document.getElementById('row-chart-foo');
      if (div) {
        document.body.removeChild(div);
      }
      jest.clearAllMocks();
    });

    it('does nothing when no data', () => {
      const target = shallow(
        <RowChart colorScheme={[]} data={[]} id="foo" title="test" total={0} />,
      );
      target._redrawChart = jest.fn();
      target.setProps({ data: [] });
      expect(target._redrawChart).toHaveBeenCalledTimes(0);
    });

    xit('trigger a new update when data changes', () => {
      const target = shallow(
        <RowChart
          colorScheme={[]}
          title="test"
          data={[
            23,
            4,
            3,
            { name: 'More Information about xy', value: 10 },
            { name: 'Athing about xy', value: 10 },
          ]}
          id="foo"
          total={1000}
        />,
      );
      target._redrawChart = jest.fn();
      const sp = jest.spyOn(target.instance(), '_redrawChart');
      target.setProps({ data: [2, 5] });
      expect(sp).toHaveBeenCalledTimes(1);
    });

    xit('trigger a new update when isPrintMode changes', () => {
      const target = shallow(
        <RowChart
          colorScheme={[]}
          title="test"
          data={[
            { name: 'fOO', value: 1 },
            { name: 'fO1', value: 1 },
            {
              name: 'Visualize trends for fOO',
              value: 1,
            },
          ]}
          id="foo"
          total={1000}
          isPrintMode={false}
        />,
      );
      target._redrawChart = jest.fn();
      const sp = jest.spyOn(target.instance(), '_redrawChart');
      target.setProps({ isPrintMode: true });
      expect(sp).toHaveBeenCalledTimes(1);
    });

    xit('trigger a new update when width changes', () => {
      const target = shallow(
        <RowChart
          colorScheme={[]}
          title="test"
          data={[23, 4, 3]}
          id="foo"
          total={1000}
          isPrintMode={false}
          width={1000}
        />,
      );
      target._redrawChart = jest.fn();
      const sp = jest.spyOn(target.instance(), '_redrawChart');
      target.setProps({ width: 600 });
      expect(sp).toHaveBeenCalledTimes(1);
    });

    xit('calls select Focus with a lens', () => {
      const cb = jest.fn();
      const target = shallow(
        <RowChart
          aggs={{ product: [1, 2, 3] }}
          selectFocus={cb}
          colorScheme={[]}
          title="test"
          lens="Product"
          data={[23, 4, 3]}
          id="foo"
          total={1000}
        />,
      );
      target.instance()._selectFocus({ name: 'foo' });
      expect(cb).toHaveBeenCalledTimes(1);
      expect(cb).toHaveBeenCalledWith({ name: 'foo' }, 'Product', [1, 2, 3]);
    });

    xit('calls select Focus with product lens - Overview', () => {
      const cb = jest.fn();
      const target = shallow(
        <RowChart
          aggs={{ product: [1, 2, 3] }}
          lens="Overview"
          selectFocus={cb}
          colorScheme={[]}
          title="test"
          data={[23, 4, 3]}
          id="foo"
          total={1000}
        />,
      );
      target.instance()._selectFocus({ name: 'foo' });
      expect(cb).toHaveBeenCalledTimes(1);
      expect(cb).toHaveBeenCalledWith({ name: 'foo' }, 'Product', [1, 2, 3]);
    });

    describe('row toggles', () => {
      let expandCb, collapseCb;
      beforeEach(() => {
        collapseCb = jest.fn();
        expandCb = jest.fn();
      });

      xit('collapses a row', () => {
        const target = shallow(
          <RowChart
            lens="Overview"
            collapseRow={collapseCb}
            expandRow={expandCb}
            colorScheme={[]}
            title="test"
            data={[{ name: 'a', isParent: true }]}
            expandedRows={['a']}
            id="foo"
            total={1000}
          />,
        );
        target.instance()._toggleRow('a');
        expect(collapseCb).toHaveBeenCalledTimes(1);
        expect(collapseCb).toHaveBeenCalledWith('a');
        expect(expandCb).toHaveBeenCalledTimes(0);
      });

      xit('expands a row', () => {
        const target = shallow(
          <RowChart
            lens="Overview"
            collapseRow={collapseCb}
            expandRow={expandCb}
            colorScheme={[]}
            title="test"
            data={[{ name: 'a', isParent: true }]}
            expandedRows={[]}
            id="foo"
            total={1000}
          />,
        );
        target.instance()._toggleRow('a');
        expect(expandCb).toHaveBeenCalledTimes(1);
        expect(expandCb).toHaveBeenCalledWith('a');
        expect(collapseCb).toHaveBeenCalledTimes(0);
      });

      xit('ignores non-parent rows', () => {
        const target = shallow(
          <RowChart
            lens="Overview"
            collapseRow={collapseCb}
            expandRow={expandCb}
            colorScheme={[]}
            title="test"
            data={[{ name: 'a', isParent: false }]}
            expandedRows={[]}
            id="foo"
            total={1000}
          />,
        );
        target.instance()._toggleRow('a');
        expect(expandCb).toHaveBeenCalledTimes(0);
        expect(collapseCb).toHaveBeenCalledTimes(0);
      });
    });
  });

  xdescribe('mapDispatchToProps', () => {
    let dispatch, gaSpy;
    beforeEach(() => {
      dispatch = jest.fn();
      gaSpy = jest.spyOn(utils, 'sendAnalyticsEvent');
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    it('hooks into changeFocus', () => {
      jest.spyOn(trendsUtils, 'scrollToFocus');
      const filters = [
        {
          key: 'A',
          'sub_product.raw': {
            buckets: [{ key: 'B' }, { key: 'C' }, { key: 'D' }, { key: 'E' }],
          },
        },
        {
          key: 'Debt collection',
          'sub_product.raw': {
            buckets: [
              { key: 'Other debt' },
              { key: 'Credit card debt' },
              { key: 'I do not know' },
              { key: 'Medical debt' },
              { key: 'Auto debt' },
              { key: 'Payday loan debt' },
            ],
          },
        },
      ];

      const element = {
        name: 'Visualize trends for A',
        parent: 'A',
      };
      mapDispatchToProps(dispatch).selectFocus(element, 'Product', filters);
      expect(dispatch.mock.lastCall).toEqual([
        [
          {
            payload: {
              filterValues: ['A', 'A•B', 'A•C', 'A•D', 'A•E'],
              focus: 'A',
              lens: 'Product',
            },
            meta: {
              requery: 'REQUERY_ALWAYS',
            },
            type: 'query/changeFocus',
          },
        ],
      ]);
      expect(trendsUtils.scrollToFocus).toHaveBeenCalled();
      expect(gaSpy).toHaveBeenCalledWith('Trends click', 'A');
    });

    it('hooks into changeFocus - no filter found', () => {
      jest.spyOn(trendsUtils, 'scrollToFocus');
      const filters = [
        {
          key: 'Debt collection',
          'sub_product.raw': {
            buckets: [
              { key: 'Other debt' },
              { key: 'Credit card debt' },
              { key: 'I do not know' },
              { key: 'Medical debt' },
              { key: 'Auto debt' },
              { key: 'Payday loan debt' },
            ],
          },
        },
      ];

      const element = {
        name: 'Visualize trends for A',
        parent: 'A',
      };
      mapDispatchToProps(dispatch).selectFocus(element, 'Product', filters);
      expect(dispatch.mock.lastCall).toEqual([
        [
          {
            payload: {
              filterValues: [],
              focus: 'A',
              lens: 'Product',
            },
            meta: {
              requery: 'REQUERY_ALWAYS',
            },
            type: 'query/changeFocus',
          },
        ],
      ]);
      expect(trendsUtils.scrollToFocus).toHaveBeenCalled();
      expect(gaSpy).toHaveBeenCalledWith('Trends click', 'A');
    });

    it('hooks into changeFocus - Company', () => {
      jest.spyOn(trendsUtils, 'scrollToFocus');
      const filters = [{ key: 'Acme' }, { key: 'Beta' }];

      const element = {
        name: 'Visualize trends for Acme',
        parent: 'Acme',
      };
      mapDispatchToProps(dispatch).selectFocus(element, 'Company', filters);
      expect(dispatch.mock.lastCall).toEqual([
        [
          {
            payload: {
              filterValues: ['Acme'],
              focus: 'Acme',
              lens: 'Company',
            },
            meta: {
              requery: 'REQUERY_ALWAYS',
            },
            type: 'query/changeFocus',
          },
        ],
      ]);
      expect(trendsUtils.scrollToFocus).toHaveBeenCalled();
      expect(gaSpy).toHaveBeenCalledWith('Trends click', 'Acme');
    });

    it('hooks into collapseTrend', () => {
      jest.spyOn(trendsUtils, 'scrollToFocus');
      mapDispatchToProps(dispatch).collapseRow('Some Expanded row');
      expect(dispatch.mock.calls).toEqual([
        [
          {
            meta: {
              requery: 'REQUERY_NEVER',
            },
            type: 'view/collapseRow',
            payload: 'Some Expanded row',
          },
        ],
      ]);
      expect(trendsUtils.scrollToFocus).not.toHaveBeenCalled();
      expect(gaSpy).toHaveBeenCalledWith(
        'Bar chart collapsed',
        'Some Expanded row',
      );
    });

    it('hooks into expandTrend', () => {
      jest.spyOn(trendsUtils, 'scrollToFocus');
      mapDispatchToProps(dispatch).expandRow('collapse row name');
      expect(dispatch.mock.calls).toEqual([
        [
          {
            meta: {
              requery: 'REQUERY_NEVER',
            },
            type: 'view/expandRow',
            payload: 'collapse row name',
          },
        ],
      ]);
      expect(trendsUtils.scrollToFocus).not.toHaveBeenCalled();
      expect(gaSpy).toHaveBeenCalledWith(
        'Bar chart expanded',
        'collapse row name',
      );
    });
  });

  describe('mapStateToProps', () => {
    let state;
    beforeEach(() => {
      state = {
        map: {},
        query: {
          lens: 'Foo',
          tab: 'Map',
        },
        trends: {},
        view: {
          expandedRows: [],
          isPrintMode: false,
          width: 1000,
        },
      };
    });
    it('maps state and props - Map', () => {
      const ownProps = {
        id: 'baz',
      };
      const actual = mapStateToProps(state, ownProps);
      expect(actual).toEqual({
        expandedRows: [],
        lens: 'Product',
        isPrintMode: false,
        tab: 'Map',
        width: 1000,
      });
    });

    it('maps state and props - Other', () => {
      const ownProps = {
        id: 'baz',
      };

      state.query.tab = 'Trends';

      const actual = mapStateToProps(state, ownProps);
      expect(actual).toEqual({
        expandedRows: [],
        lens: 'Foo',
        isPrintMode: false,
        tab: 'Trends',
        width: 1000,
      });
    });
  });

  describe('helper functions', () => {
    xit('gets height based on number of rows', () => {
      const target = mount(
        <RowChart
          colorScheme={[]}
          title="test"
          total={10}
          data={[23, 4, 3]}
          id="foo"
        />,
      );
      let res = target.instance()._getHeight(1);
      expect(res).toEqual(100);
      res = target.instance()._getHeight(5);
      expect(res).toEqual(300);
    });

    xit('formats text of the tooltip', () => {
      const target = mount(
        <RowChart
          colorScheme={[]}
          title="test"
          total={1000}
          data={[23, 4, 3]}
          id="foo"
        />,
      );
      const res = target.instance()._formatTip(100000);
      expect(res).toEqual('100,000 complaints');
    });
  });
});
