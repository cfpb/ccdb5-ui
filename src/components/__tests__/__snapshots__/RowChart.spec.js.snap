// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: RowChart initial state renders without crashing 1`] = `
<div
  className="row-chart-section"
>
  <h3>
    Foo title we want
  </h3>
  <p>
    sub title under title
  </p>
  <div
    id="row-chart-foo"
  />
</div>
`;
